### PR TITLE
preflight check enhancement

### PR DESCRIPTION
**Pull request by Simonzhaohui**
Features:
  - Create SPN with MS Graph API permissions:
  - Create SPN with subscription owner
  - Assign SPN and Signed user as App and SPN owner
  - Grant admin consent

BUG FIXES:
  - Update validate user check

### CODE REVIEW

**Review by xugy99 — APPROVED** [review]
+1 with the assumption of the subject and description of this PR is changed  to reflect the changes in the code

**Review by xugy99 — CHANGES_REQUESTED** [review]
I am a bit regret on the +1 as the  `create-spn.sh` carries more things than its name indicates: the name says "create" but it can also be used to "update with more permission by graph API".

**Comment by Simonzhaohui** [comment]
@xugy99 Rename `create-spn.sh` to `az-spn-generator.sh` ?

File: azure/create-spn.sh [hunk 1]
```diff
@@ -135,24 +150,70 @@ show_account()
...
+    --headers Content-Type=application/json \
+    -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/servicePrincipals/${sp_object_id}\"}"
+  az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/${sp_object_id}/owners/\$ref \
+    --headers Content-Type=application/json \
+    -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/users/${user_object_id}\"}"
+
+  # To add API permissions
+  # Application.ReadWrite.OwnedBy
+  az ad app permission add --id "${client_id}" --only-show-errors \
+    --api 00000003-0000-0000-c000-000000000000 \
```
Comment (nstott) [inline azure/create-spn.sh:187]: these ids are magical, 
can we add some comments/information about what these magic ids mean

File: azure/create-spn.sh [hunk 2]
```diff
@@ -135,24 +150,70 @@ show_account()
...
 create_ad_sp()
 {
-  echo "Creating Azure AD Service Principal and configuring its access to Azure resources in subscription ${subscription}..."
-  years="${years:-1}"
-  spn=$(az ad sp create-for-rbac -o json -n ${display_name} --role Owner --scopes /subscriptions/${subscription} --years ${years})
+  if [[ -z "${client_id}" ]]; then
+    echo "Creating Azure AD Service Principal and configuring its access to Azure resources in subscription ${subscription}..."
+    years="${years:-1}"
+    spn=$(az ad sp create-for-rbac -o json -n "${display_name}" --only-show-errors \
+      --role Owner --scopes /subscriptions/${subscription} --years ${years})
```
Comment (nstott) [inline azure/create-spn.sh:157]: we're inconsistent with the quotes, 
we should prefer "${subscription}" instead of ${subscription}

**Review by nstott — APPROVED** [review]
lgtm, 
some small nits inline, 
we should prefer "${thing}" instead of ${thing} or $thing

thanks for getting this in!

File: azure/create-spn.sh (hunk 1 above)
Comment (Simonzhaohui) [reply azure/create-spn.sh:218]: add comments for API permissions.

File: azure/create-spn.sh (hunk 2 above)
Comment (Simonzhaohui) [reply azure/create-spn.sh]: Updated

**Review by xugy99 — APPROVED** [review]

**Pull request merged**